ratchet-tree.go

```diff
@@ -0,0 +1,479 @@
+package mls
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/bifurcation/mint/syntax"
+)
+
+//// Tree Hashes
+type ParentNodeInfo struct {
+	PublicKey      *HPKEPublicKey
+	UnmergedLeaves []leafIndex `tls:"head=4"`
+}
+
+type ParentNodeHashInput struct {
+	HashType  uint8
+	Info      *ParentNodeInfo `tls:"optional"`
+	LeftHash  []byte          `tls:"head=1"`
+	RightHash []byte          `tls:"head=1"`
+}
+
+type LeafNodeInfo struct {
+	PublicKey  HPKEPublicKey
+	Credential Credential
+}
+
+type LeafNodeHashInput struct {
+	HashType uint8
+	Info     *LeafNodeInfo `tls:"optional"`
+}
+
+//// ratchet tree node and helpers
+type OptionalRatchetNode struct {
+	Node *RatchetTreeNode
+	Hash []byte `tls:"head=2"`
+}
+
+func (n *OptionalRatchetNode) setLeafHash(cs CipherSuite) {
```

Comment:
Rather than passing the CipherSuite ll the time, just store it on the node, with `tls:"omit"`.